internal/backend/sftp/sftp.go

```diff
@@ -155,15 +155,29 @@ func Open(ctx context.Context, cfg Config) (*SFTP, error) {
 	return sftp, nil
 }
 
-func (r *SFTP) mkdirAllDataSubdirs() error {
+func (r *SFTP) mkdirAllDataSubdirs(ctx context.Context, nconn uint) error {
+	// Run multiple MkdirAll calls concurrently. These involve multiple
+	// round-trips and we do a lot of them, so this whole operation can be slow
+	// on high-latency links.
+	g, _ := errgroup.WithContext(ctx)
+	// Use errgroup's built-in semaphore, because r.sem is not initialized yet.
+	g.SetLimit(int(nconn))
+
 	for _, d := range r.Paths() {
-		err := r.c.MkdirAll(d)
-		if err != nil {
-			return err
-		}
+		d := d
+		g.Go(func() error {
+			// First try Mkdir. For most directories in Paths, this takes one
+			// round trip, not counting duplicate parent creations causes by
+			// concurrency. MkdirAll first does Stat, then recursive MkdirAll
+			// on the parent, so calls typically take three round trips.
+			if err := r.c.Mkdir(d); err == nil {
+				return nil
+			}
+			return r.c.MkdirAll(d)
```

Comment:
comment only, no change suggested:

There is a scenario where we concurrently attempt to make two or more child directories e.g. `somerepo/data/00` and `somerepo/data/01` before the parent `somerepo/data` has been created. 

I was a bit concerned about this, that it could lead to failures in some (rare? cases), but after reading through the pkg/sftp code and thinking about it, it is probably fine:

Suppose goroutine A will process `somerepo/data/00`, goroutine B will process `somerepo/data/01` and goroutine C will process `somerepo/data` ... assume C will be scheduled later, and we think about A and B.

If `somerepo/data/00` and `somerepo/data/01` are processed sequentially, there is no problem.

Suppose execution of `somerepo/data/00` `somerepo/data/01` is interleaved.

A tricky interleaving could be where both A and B both decide that the parent directory `somerepo/data` does not exist, then try to create it at the same time:

```
A Mkdir : `SSH_FXP_MKDIR` "somerepo/data/00" # error as parent dir data doesn't exist
B Mkdir : `SSH_FXP_MKDIR` "somerepo/data/01" # error as parent dir data doesn't exist
A MkdirAll: `SSH_FXP_STAT` "somerepo/data/00 # result: this child dir doesn't exist
B MkdirAll: `SSH_FXP_STAT` "somerepo/data/01 # result: this child dir doesn't exist

# both MkdirAll calls recursively attempt to create the parent dir "somerepo/data"
# first they stat the path

A MkdirAll : `SSH_FXP_STAT` "somerepo/data" # result: this parent dir doesn't exist
B MkdirAll : `SSH_FXP_STAT` "somerepo/data" # result: this parent dir doesn't exist

# then they both try to create it:

A MkdirAll : `SSH_FXP_MKDIR` "somerepo/data" # suppose this is processed by server first and succeeds
B MkdirAll : `SSH_FXP_MKDIR` "somerepo/data" # fails, cannot mkdir since data directory already exists
```

So there is some interleaving where both goroutines A and B attempt to create the same parent `data` dir, and the operation that is processed second fails.

I suspect this may be okay in practice as there is code in the `pkg/sftp` client implementation of `MkdirAll` that does an `SSH_FXP_LSTAT` operation if there is an error following `SSH_FXP_MKDIR`.  If the LSTAT succeeds and says that the path exists and is a dir, then it looks like the client code for `MkdirAll` will ignore the error from failing to create a dir that already exists.

https://github.com/pkg/sftp/blob/v1.13.4/client.go#L899-L908

Reply:
sftp.Client.MkdirAll actually promises that this will always work:

    // If path is already a directory, MkdirAll does nothing and returns nil.